compress/lz4/lz4.go

```diff
@@ -41,115 +37,44 @@ func (c *Codec) CompressionCodec() format.CompressionCodec {
 	return format.Lz4Raw
 }
 
-func (c *Codec) NewReader(r io.Reader) (compress.Reader, error) {
-	return &reader{reader: r}, nil
-}
-
-func (c *Codec) NewWriter(w io.Writer) (compress.Writer, error) {
-	return &writer{writer: w, compressor: lz4.CompressorHC{Level: c.Level}}, nil
-}
-
-type reader struct {
-	buffer bytes.Buffer
-	data   []byte
-	offset int
-	reader io.Reader
-}
-
-func (r *reader) Close() error {
-	r.offset = len(r.data)
-	r.reader = nil
-	return nil
-}
+func (c *Codec) Encode(dst, src []byte) ([]byte, error) {
+	dst = reserveAtLeast(dst, len(src)/4)
 
-func (r *reader) Reset(rr io.Reader) error {
-	r.buffer.Reset()
-	r.data = r.data[:0]
-	r.offset = 0
-	r.reader = rr
-	return nil
-}
-
-func (r *reader) Read(b []byte) (n int, err error) {
-	if r.offset == 0 && len(r.data) == 0 {
-		if err := r.decompress(); err != nil {
-			return 0, err
+	compressor := lz4.CompressorHC{Level: c.Level}
+	for {
+		n, err := compressor.CompressBlock(src, dst)
+		if err != nil { // see Decode for details about error handling
+			dst = make([]byte, 2*len(dst))
+		} else {
+			return dst[:n], nil
 		}
 	}
-	n = copy(b, r.data[r.offset:])
-	r.offset += n
-	if r.offset == len(r.data) {
-		err = io.EOF
-	}
-	return n, err
 }
 
-func (r *reader) decompress() error {
-	if r.reader == nil {
-		return io.EOF
-	}
-
-	_, err := r.buffer.ReadFrom(r.reader)
-	if err != nil {
-		return err
-	}
-
-	optimisticOutputSize := 3 * r.buffer.Len()
-	if cap(r.data) < optimisticOutputSize {
-		r.data = make([]byte, optimisticOutputSize)
-	} else {
-		r.data = r.data[:cap(r.data)]
-	}
+func (c *Codec) Decode(dst, src []byte) ([]byte, error) {
+	dst = reserveAtLeast(dst, 3*len(src))
```

Comment:
I guess I'd put a comment in here, like, "observation shows that decoded bytes are 3x the size of encoded bytes" or something, or just, an answer to the question "how did you pick the number three"